### PR TITLE
Set up channel receivers early

### PR DESCRIPTION
**Pull request by lupine**
Problem identified while working on #143 ; this is my attempt at a fix. Another option exists in #144, and there may be other, better, ways too.

At present, components can race with each other and try to send a message on a channel before the other side has managed to set up their receiver. When this happens, an error is raised, which can lead to process termination.

Instead, set up all the receivers at the initialization stage (e.g. when calling new()). This requires some trickery to pass the mutable receiver down into the respective start() -> receive() methods, which is achieved with an option inside a cell. When the mutable receiver is consumed, it is replaced with a None value.

The stop methods of the various structs that need this all consume the reference (i.e. take &mut self), so there's no chance of the start() method being called a second time (leading to the `expect` calls being triggered).

### CODE REVIEW

File: src/coordinator/mod.rs [hunk 1]
```diff
@@ -14,11 +15,25 @@ pub type InputsStore = std::collections::HashMap<Serial, lxp::packet::ReadInputs
 pub struct Coordinator {
     config: ConfigWrapper,
     channels: Channels,
+
+    inverter_receiver: std::cell::Cell<Option<inverter::Receiver>>,
```
Comment (lupine) [inline src/coordinator/mod.rs:19]: The cell gives us interior mutability; the option gives us the ability to swap out the mutable receiver and put a `None` in its place instead.

There might be better ways to achieve the same thing; the critical aspect of this design is that the receiver is created in the initializer, rather than as a result of the `start()` method.

File: src/mqtt.rs [hunk 2]
```diff
@@ -231,19 +231,27 @@ pub enum ChannelData {
...
     config: ConfigWrapper,
     shutdown: bool,
     channels: Channels,
+    receiver: std::cell::Cell<Option<Receiver>>,
 }
 
 impl Mqtt {
     pub fn new(config: ConfigWrapper, channels: Channels) -> Self {
+        // Create the receiver at new() time, rather than start(), so things
+        // wanting to send to MQTT queue messages, rather than failing with an
```
Comment (lupine) [inline src/mqtt.rs:247]: I was surprised to learn that these tokio broadcast channels have a fixed-size buffer (set to 512 in src/channels.rs) and that messages are dropped when the buffer is full, rather than send() async-blocking until a recv() call happened on the other side.

This design works with either approach, but where it matters most is when MQTT is disabled, where if it were the latter case, we'd have to do extra work to avoid blocking forever.

As-is, rather than the message being discarded with an error (there are examples in the codebase of that error being picked up on), the send() will now succeed. 512 messages will be kept in RAM.

It would be better for the sender to notice that MQTT is disabled and not try to send the message to begin with. I can add that to this PR if desired.

I guess a similar situation prevails with influx?

File: src/coordinator/mod.rs [hunk 3]
```diff
@@ -14,11 +15,25 @@ pub type InputsStore = std::collections::HashMap<Serial, lxp::packet::ReadInputs
...
+    mqtt_receiver: std::cell::Cell<Option<mqtt::Receiver>>,
 }
 
 impl Coordinator {
     pub fn new(config: ConfigWrapper, channels: Channels) -> Self {
-        Self { config, channels }
+        // Create the receivers at new() time, rather than start(), so things
+        // wanting to send to us will queue messages instead of failing with an
+        // error about closed channels
+        let inverter_receiver = std::cell::Cell::new(Some(channels.from_inverter.subscribe()));
```
Comment (lupine) [inline src/coordinator/mod.rs:28]: The coordinator's commands also have receivers on `from_inverter`, which are created as needed in the `run` methods rather than at initialization. I've left those alone, since they don't seem to suffer from the same kind of problem.

**Comment by lupine** [comment]
I've tested this manually by rebasing #143 against it, removing the sleep, and running it against my inverter. It sends the messages on startup and everything else looks to be operating as normal as well.

I've not tried the without-mqtt-enabled or without-influx-enabled cases, but I'm pretty confident in the dead reckoning above.

Let me know what you think!

File: src/lxp/inverter.rs [hunk 4]
```diff
@@ -130,17 +130,24 @@ pub struct Inverter {
...
 
 impl Inverter {
     pub fn new(config: ConfigWrapper, inverter: &config::Inverter, channels: Channels) -> Self {
         // remember which inverter this instance is for
         let host = inverter.host().to_string();
 
+        // Create the receiver at new() time, rather than start(), so things
+        // wanting to send to the inverter queue messages, rather than failing
+        // with an error about closed channels
+        let receiver = std::cell::Cell::new(Some(channels.to_inverter.subscribe()));
```
Comment (lupine) [inline src/lxp/inverter.rs:144]: FWIW I was quite surprised to discover there is only one channel for all inverters. In my head, it was going to be a channel per inverter.

I don't know if it makes a practical difference for this PR, but it was giving me all kinds of grief and edge cases when fiddling with `receiver_count`.

File: tests/test_inverter.rs [hunk 5]
```diff
@@ -112,8 +112,6 @@ async fn test_replies_to_heartbeats() {
     let channels = Channels::new();
     let inverter = lxp::inverter::Inverter::new(config, &inverter, channels.clone());
 
-    let from_inverter = channels.from_inverter.subscribe();
```
Comment (lupine) [inline tests/test_inverter.rs:115]: Busted :D. I'm guessing this was here to prevent the error message about closed channels from occurring.

With this change, the `Inverter::new` call on line 113 creates a subscription for us, so we don't need this line here. It resolves that clippy lint too.

Comment (celsworth) [reply tests/test_inverter.rs:115]: Yeah clippy has been moaning at me about this one for a while, nice to get rid of it :)

File: src/lxp/inverter.rs (hunk 4 above)
Comment (celsworth) [reply src/lxp/inverter.rs:144]: I've wondered about changing this before as I agree its surprising but there's never been a compelling case to do so. So it would have been change for the sake of it.

I wouldn't be opposed to changing that if it doesn't add too much extra complication.

File: src/coordinator/mod.rs (hunk 1 above)
Comment (celsworth) [reply src/coordinator/mod.rs:19]: Bit of a shame about the added complexity. I wonder if there's a case for merging `new` and `start`? I'm not sure if there's any reason we couldn't really. So just creating an instance is enough to get things running; maybe it would be a bit weird for `new` to return a future though..

File: src/mqtt.rs (hunk 2 above)
Comment (celsworth) [reply src/mqtt.rs:247]: 512 was a somewhat arbitrary choice that I just hoped we'd never run into :) Its enough for all the input registers to be queued and then we pretty much just hope they get cleared before more come in. In normal circumstances they're gone in a second or so.

I thought most places that tried to send to MQTT already checked via something like `self.config.mqtt().enabled()` - if any don't then that should probably be added.

This setup all starts to get a bit messy when things aren't behaving, I'm not sure what the behaviour might be if we async-block on sending to a channel anywhere? If the coordinator is blocked then we'll probably run into all sorts of other issues. Maybe we're reaching the limitations of this design. I'm pretty open to suggestions on improvements here, the use of channels in this way was a very early decision while learning rudimentary Rust :)

Also looks like the `broadcast` channels currently in use don't support waiting to send, we'd have to swap to `mpsc` to get that: https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Sender.html#method.send - but I don't think that would be a problem. This one also supports unbounded channels so we could keep queueing forever but that feels like a memory balloon waiting to happen. Would be better to deterministically decide what happens if things aren't working and messages aren't being consumed.

File: src/coordinator/mod.rs (hunk 1 above)
Comment (lupine) [reply src/coordinator/mod.rs]: OK, I've put `Cell` in the prelude.

Merging `new` and `start` would bring back the same problem - we'd be creating receivers in `newstart` in some components while other components would be trying to send to them in their own `newstart`s. Moving the receiver creation to `new` is what makes this work :sweat_smile:

I think using `Cell` is quite normal in this case, but the `Option` is a bit yucky, for sure. It has the benefit of being a small change with the existing code.

It would be a bigger refactor, but we could get rid of the `Option`s by moving the code in the respective `stop()` methods to a drop trait (or otherwise doing away with it), then changing `start` to be `&mut self`. We'd then be able to use `Cell.into_inner()` instead of `replace()` to access the receiver, I think.

File: src/lxp/inverter.rs (hunk 4 above)
Comment (lupine) [reply src/lxp/inverter.rs]: I'd rather leave it out of this PR if that's alright with you :sweat_smile:. If we were going the `receiver_count` route I'd definitely have been doing it, but this way it's neither here nor there, I think.

File: src/mqtt.rs (hunk 2 above)
Comment (lupine) [reply src/mqtt.rs:247]: I was just chewing the fat here TBH. My own rust is pretty early-stage, but the general idea seems to be to design these things out through the type system, as with the idea about removing `Option` above.

Do you want us to make further progress on this before merging this PR, or are you OK with it going in as-is?

If we need further changes, one thought I had was a null object sort of thing - basically a discard sink that could be the subscriber to the `to_mqtt` channel when mqtt is disabled. Something similar for influx and the other disable-able components too. That would allow the `if ... enabled()` conditionals to go away, along with the `bail!`s around the closed channel messages that are currently here. That could go into this PR if you don't like the buffers filling up when mqtt is disabled? Just let me know.

Silently dropping messages when the buffer is full is :grimacing:,  and switching to unbuffered, or at least blocks-when-full, channels fixes that, but it has all sorts of implications for the design of the rest of the system to keep it from deadlocking anything :sweat_smile:. It's a thought for how to go in the future, but I'd hate for the holding register publish stuff to be be (ahahaha) blocked behind

Comment (lupine) [reply src/mqtt.rs:247]: Ah, the `bail` I had in mind is in `src/coordinator/commands/time_register_ops.rs`, so would only ever be called if mqtt were enabled anyway (although that's not guaranteed by construction, it happens to be true).

The only way it (and the others in `mod.rs`) could fire now with mqtt enabled is during shutdown, I think - this PR designs out the possibility that they could fire while mqtt is enabled and `stop` hasn't been called. Could probably design that out too, as part of making the `stop` methods go away.

File: src/coordinator/mod.rs (hunk 1 above)
Comment (celsworth) [reply src/coordinator/mod.rs]: Its a very minor thing, but I see `Cell` has a `take` that replaces the value with `default`, which would be `None`: https://doc.rust-lang.org/stable/std/option/enum.Option.html#impl-Default-for-Option%3CT%3E

Think it'd be a nice little cleanup to use that instead.

After that I'm happy to merge this.

Comment (lupine) [reply src/coordinator/mod.rs]: Great :thumbsup: - I saw it, but before I'd realised I needed the `Option` :sweat_smile:. I've made that change now.
